### ISSUE TITLE
Question : where to to start

### ISSUE THREAD

**Issue by interstar** [d-cook/Objects#12]
Hi. d-cook

This looks interesting. But what's the relation between this project and your Bootstrap lab?

Is one using the other? Are they forks with two copies of the same thing? Are they different?

I'm kind of looking for the easiest to understand and work-with toy object environment (ie. a place for making objects which can send messages to each other) because I want to try making a toy example of what I'm describing here : http://sdi.thoughtstorms.info/?p=1344

Basically I want to make a simple concept demonstrator of an "assemblage language" ie. something that compiles down to a bunch of objects living together in a world.

Would this project be suitable as a target for that?

Do you have a really easy, idiot's guide to show how to create some new objects, populated with data and living in a live environment?

**Comment by d-cook** [comment]
[Bootstrap lab](https://github.com/jdjakub/BootstrapLab) is a separate project by Joel. However, both are trying to solve very similar problems in a very similar fashion (or at least, in a fashion that builds upon very similar principles).

The similarity is a huge coincidence, because we arrived at these things completely separately. But it's also not, because the fact that we recognized the same issues, drew the same conclusions, and even have very similar solutions, is a strong affirmation that what we have recognized and seek to fix is very real. Joel actually has a great [write-up about this](https://programmingmadecomplicated.wordpress.com/2017/08/13/language-isnt-everything/) on his blog

**Comment by d-cook** [comment]
What I have here is a work in progress, and there's not much to show for it (not visually, anyway).

My idea is a tool / platform for direct manipulate of program entities at runtime, including the entities that make up the tool itself, so that you can reshape it into anything you want. There are two parts required to make this work: an underlying model/engine that makes it all work, and GUI through which to do it all.

THIS project starts with the underlying model, and is far from being a visual tool (though that is the ultimate plan). You might actually be more interested in [https://github.com/d-cook/Interact](Interact), which starts with the UI and will work it's way down to bootstrapping itself. That is a work-in-progress too with not much to show yet, but it at least has *something* to interact with. Also, it does have a functioning language & interpreter in it already.

As for if it's a suitable tool for what you're after, the idea is for it to be moldable into the ideal tool for *anything* ... at least in theory / eventually. But if you are wanting something that is open-ended from top-to-bottom, then you need a tool like this.

You might also be interested in my [https://github.com/d-cook/Render](Render] project, which makes it VERY quick and easy to dynamically render content. You must create an instance of it (`var r = Render(...);`) and add its canvas (`r.getCanvas()`) to the HTML document; but after that, you just call `r.render([...])` with the full spec of what to render, and it does the rest. See the readme there for more info.

**Comment by d-cook** [comment]
I'll have to read your article ("[What's wrong with OOP?](http://sdi.thoughtstorms.info/?p=1344)") before I provide a more specific answer for that, but I suspect we'll have some good material to discuss (whether or not what I'm doing fits your needs).

**Comment by d-cook** [comment]
If you're looking for something built specifically to support the kind of "assemblage" you have in mind, that's not what I have here.

But if you are just looking for a tool for ad-hoc manipulation and visualization of objects, then this might work for you (when it ever gets far enough). But again, my other projects (Interact or Render) are probably better for that.

**Comment by d-cook** [comment]
Your assemblage idea reminds me a lot about Jim Coplien's DCI paradigm. It focuses on reflecting the networks of cooperating objects explicit in source code. The DCI pattern defines a Context (very much like your assemblage) which defines a certain kind of interaction relationship between a network of objects. The objects in a Context are identified by the role that they play in the interaction, and each role can have methods, which can interact with the other objects (in terms of role).

For example, a restaurant context has a customer, a waiter, a chef, etc and methods like order food, take order, cook food, deliver food, pay bill, etc. These things are not intrinsic to the class called "person", but any person (or anything that fits the interface) may play any role. Even if objects change roles (e.g. now the waiter is having lunch, or shifts change and someone else finishes waiting on a table), the interactions and roles do not change, so DCI allows the code to capture the interactions between roles within a context as a static structure in the code.

Maybe it's not quite the same thing you talk about, but I think there's some similar underlying principles and goals.

Example videos:
* https://www.infoq.com/presentations/Reflection-OOP-Social
* https://youtu.be/SxHqhDT9WGI
